gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java

```diff
@@ -382,35 +344,49 @@ private Connection waitForConnection(final long timeout, final TimeUnit unit) th
         long start = System.nanoTime();
         long remaining = timeout;
         long to = timeout;
-        do {
-            try {
-                awaitAvailableConnection(remaining, unit);
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                to = 0;
+
+        Connection leastUsed = selectLeastUsed();
```

Comment:
Perhaps refactor as following for returning the happy case early. It helps in readability since the reader now understands that the code after this line is handling logic for case when no connection is immediately available.

You can do something like:
```
// scenario 1: when connection pool is empty
if (connections.empty()) {
    for (size 1..minPoolSize)
        createNewConnection()
    return waitForConnection()
}

// scenario 2: happy case when a connection is available
Connection leastUsed = selectLeastUsed();
if (leastUsed != null && leastUsed.canAcceptNewRequests)
    return leastUsed

// scenario 3: wait for a connection to become available
maybeCreateNewConnection() // checks if it is possible to create a new connection, if yes, creates one.
return waitForConnection() // doesn't need to consider the older leastUsed, it will wait first and then ask another.

fun waitForConnection {
  do {
      awaitForAvailableConnection
      find least used
      if (leastUsed.canAcceptNewRequests)
            return leastUsed
  } while (timeout pending)
}
```

Reply:
I am confused as to what is being proposed here.

This is more or less the way the existing logic is written. The ticket was created with the acceptance criteria to avoid the duplication of the checking in scenario 2 and 3, where now it is combined into a single point of checking in `waitForConnection`. 

Whilst I agree that it is now less readable, unfortunately this is the result. 

I will add some comments to make it more clear what is happening.